File: tools/serve.py
#!/usr/bin/env python3
"""Serve the page and relay JAKIM e-Solat from one origin.

The browser refuses to read e-solat.gov.my from a page on another origin
because e-Solat sends no Access-Control-Allow-Origin header. Nothing you can
write in the page fixes that -- the check is the browser's, and the header is
the server's. The way out is to stop making it a cross-origin request: serve
the schedule from the same host that serves the page.

    python3 tools/serve.py
    # then open http://localhost:8000
    # and set the relay field to:  /api/jakim?zone={zone}&period={period}

Standard library only, same as `python3 -m http.server`.

This is not a general proxy. The zone and period are validated against fixed
patterns and the upstream URL is rebuilt from them, so it can only ever fetch
prayer times -- it cannot be pointed at anything else by a crafted request.
"""

import argparse
import json
import os
import re
import sys
import urllib.error
import urllib.parse
import urllib.request
from functools import partial
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer

DEFAULT_UPSTREAM = "https://www.e-solat.gov.my/index.php?r=esolatApi/takwimsolat"

ZONE_RE = re.compile(r"^[A-Z]{3}[0-9]{2}$")
PERIODS = {"today", "week", "month", "year", "duration"}

RELAY_PATH = "/api/jakim"
UPSTREAM_TIMEOUT = 30


class Handler(SimpleHTTPRequestHandler):
    upstream = DEFAULT_UPSTREAM

    def do_GET(self):
        if urllib.parse.urlparse(self.path).path == RELAY_PATH:
            self.relay()
        else:
            super().do_GET()

    def do_OPTIONS(self):
        if urllib.parse.urlparse(self.path).path == RELAY_PATH:
            self.send_response(204)
            self.cors()
            self.send_header("Access-Control-Allow-Methods", "GET, OPTIONS")
            self.end_headers()
        else:
            self.send_error(405)

    # -- relay ---------------------------------------------------------

    def relay(self):
        params = urllib.parse.parse_qs(urllib.parse.urlparse(self.path).query)
        zone = (params.get("zone") or [""])[0].strip().upper()
        period = (params.get("period") or ["year"])[0].strip().lower()

        if not ZONE_RE.match(zone):
            return self.fail(400, "kod zon tidak sah: %r" % zone)
        if period not in PERIODS:
            return self.fail(400, "period tidak sah: %r" % period)

        url = "%s&period=%s&zone=%s" % (
            self.upstream,
            urllib.parse.quote(period),
            urllib.parse.quote(zone),
        )

        req = urllib.request.Request(url, headers={
            "Accept": "application/json",
            # e-Solat has been known to answer differently without one.
            "User-Agent": "waktu-solat-relay/1.0",
        })

        try:
            with urllib.request.urlopen(req, timeout=UPSTREAM_TIMEOUT) as res:
                body = res.read()
        except urllib.error.HTTPError as e:
            return self.fail(502, "e-Solat menjawab HTTP %d" % e.code)
        except Exception as e:  # timeout, DNS, TLS, refused
            return self.fail(504, "tidak dapat menghubungi e-Solat: %s" % e)

        # Pass the bytes through untouched; the page does its own validation.
        self.send_response(200)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Content-Length", str(len(body)))
        self.cors()
        self.end_headers()
        self.wfile.write(body)
        self.log_message('relay %s period=%s -> %d bytes', zone, period, len(body))

    def fail(self, code, message):
        body = json.dumps({"status": "Relay error", "error": message}).encode("utf-8")
        self.send_response(code)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Content-Length", str(len(body)))
        self.cors()
        self.end_headers()
        self.wfile.write(body)

    def cors(self):
        # Harmless when the page is same-origin, and lets the relay also serve
        # a copy of the site hosted elsewhere.
        self.send_header("Access-Control-Allow-Origin", "*")

    # Prayer times change daily; never let a proxy or the browser pin them.
    def end_headers(self):
        if urllib.parse.urlparse(self.path).path == RELAY_PATH:
            self.send_header("Cache-Control", "no-store")
        super().end_headers()


def main():
    ap = argparse.ArgumentParser(description=__doc__,
                                 formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("-p", "--port", type=int, default=8000)
    ap.add_argument("-b", "--bind", default="0.0.0.0",
                    help="default 0.0.0.0, so other machines on the LAN can reach it")
    ap.add_argument("-d", "--directory",
                    default=os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
                    help="site root (default: the repository)")
    ap.add_argument("--upstream", default=DEFAULT_UPSTREAM,
                    help=argparse.SUPPRESS)
    args = ap.parse_args()

    Handler.upstream = args.upstream
    handler = partial(Handler, directory=args.directory)

    with ThreadingHTTPServer((args.bind, args.port), handler) as httpd:
        print("serving %s on http://%s:%d" % (args.directory, args.bind, args.port))
        print("relay at %s?zone={zone}&period={period}" % RELAY_PATH)
        print("set that path in the page's relay field, then leave this running.")
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            print("\nstopped")
            return 0


if __name__ == "__main__":
    sys.exit(main())

File: tools/serve.mjs
#!/usr/bin/env node
/*
 * Serve the page and relay the prayer-time APIs from one origin.
 *
 *   npm start
 *   # then open http://localhost:8000
 *   # and set the relay field to:  /api/solat?zone={zone}&year={year}
 *
 * Neither solat.my nor e-solat.gov.my sends an Access-Control-Allow-Origin
 * header, so a browser refuses to let a page on another origin read them. The
 * request itself is fine -- open either URL in a tab and the JSON is there --
 * but a fetch from the page is blocked. Nothing written in the page fixes
 * that: the check is the browser's and the header is the server's.
 *
 * So this stops the request being cross-origin. Node fetches upstream
 * server-side, where there is no CORS to answer to, and hands the bytes back
 * from the same origin that served the page.
 *
 * No dependencies -- Node 18+ only, for its built-in fetch.
 *
 * This is not a general proxy. The zone, year and period are validated against
 * fixed patterns and the upstream URL is rebuilt from them, so a crafted
 * request cannot point it anywhere else.
 */

import http from 'node:http';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT_DEFAULT = path.dirname(path.dirname(fileURLToPath(import.meta.url)));

const SOLATMY_BASE = 'https://solat.my/api/yearly';
const JAKIM_BASE = 'https://www.e-solat.gov.my/index.php?r=esolatApi/takwimsolat';

const ZONE_RE = /^[A-Z]{3}[0-9]{2}$/;
const PERIODS = new Set(['today', 'week', 'month', 'year', 'duration']);
const UPSTREAM_TIMEOUT_MS = 30000;

const MIME = {
  '.html': 'text/html; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.mjs': 'text/javascript; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.svg': 'image/svg+xml',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.ico': 'image/x-icon',
  '.mp3': 'audio/mpeg',
  '.ogg': 'audio/ogg',
  '.wav': 'audio/wav',
  '.txt': 'text/plain; charset=utf-8',
  '.md': 'text/markdown; charset=utf-8',
};

const args = parseArgs(process.argv.slice(2));
const ROOT = path.resolve(args.dir ?? ROOT_DEFAULT);
const PORT = Number(args.port ?? 8000);
const HOST = args.host ?? '0.0.0.0';

// Overridable so the relay can be exercised against a stand-in upstream.
const SOLATMY = args['solat-base'] ?? SOLATMY_BASE;
const JAKIM = args['jakim-base'] ?? JAKIM_BASE;

function parseArgs(argv) {
  const out = {};
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (!a.startsWith('--')) continue;
    const eq = a.indexOf('=');
    if (eq > -1) out[a.slice(2, eq)] = a.slice(eq + 1);
    else out[a.slice(2)] = argv[++i];
  }
  if (out.p) out.port = out.p;
  return out;
}

// -- relay ------------------------------------------------------------

function upstreamFor(pathname, params) {
  const zone = (params.get('zone') ?? '').trim().toUpperCase();
  if (!ZONE_RE.test(zone)) {
    return { error: `kod zon tidak sah: ${JSON.stringify(zone)}` };
  }

  if (pathname === '/api/solat') {
    const year = (params.get('year') ?? String(new Date().getFullYear())).trim();
    if (!/^[0-9]{4}$/.test(year)) {
      return { error: `tahun tidak sah: ${JSON.stringify(year)}` };
    }
    return { url: `${SOLATMY}/${encodeURIComponent(zone)}/${encodeURIComponent(year)}` };
  }

  const period = (params.get('period') ?? 'year').trim().toLowerCase();
  if (!PERIODS.has(period)) {
    return { error: `period tidak sah: ${JSON.stringify(period)}` };
  }
  return {
    url: `${JAKIM}&period=${encodeURIComponent(period)}&zone=${encodeURIComponent(zone)}`,
  };
}

async function relay(req, res, pathname, params) {
  const { url, error } = upstreamFor(pathname, params);
  if (error) return sendJSON(res, 400, { status: 'Relay error', error });

  const ac = new AbortController();
  const timer = setTimeout(() => ac.abort(), UPSTREAM_TIMEOUT_MS);

  try {
    const upstream = await fetch(url, {
      signal: ac.signal,
      headers: {
        Accept: 'application/json',
        'User-Agent': 'waktu-solat-relay/1.0',
      },
    });
    const body = Buffer.from(await upstream.arrayBuffer());

    if (!upstream.ok) {
      return sendJSON(res, 502, {
        status: 'Relay error',
        error: `hulu menjawab HTTP ${upstream.status}`,
        upstream: url,
      });
    }

    res.writeHead(200, {
      'Content-Type': 'application/json; charset=utf-8',
      'Content-Length': body.length,
      // Harmless same-origin, and lets the relay serve a copy hosted elsewhere.
      'Access-Control-Allow-Origin': '*',
      // Prayer times change daily; never let anything pin them.
      'Cache-Control': 'no-store',
    });
    res.end(body);
    log(`relay ${pathname} -> ${body.length} bytes`);
  } catch (err) {
    const why = err.name === 'AbortError' ? 'tamat masa' : err.message;
    sendJSON(res, 504, { status: 'Relay error', error: `tidak dapat menghubungi hulu: ${why}`, upstream: url });
  } finally {
    clearTimeout(timer);
  }
}

function sendJSON(res, code, obj) {
  const body = Buffer.from(JSON.stringify(obj));
  res.writeHead(code, {
    'Content-Type': 'application/json; charset=utf-8',
    'Content-Length': body.length,
    'Access-Control-Allow-Origin': '*',
    'Cache-Control': 'no-store',
  });
  res.end(body);
}

// -- static -----------------------------------------------------------

function serveStatic(req, res, pathname) {
  const rel = decodeURIComponent(pathname === '/' ? '/index.html' : pathname);
  const file = path.join(ROOT, rel);

  // path.join collapses "..", so this catches any attempt to climb out.
  if (file !== ROOT && !file.startsWith(ROOT + path.sep)) {
    res.writeHead(403).end('Forbidden');
    return;
  }

  fs.stat(file, (err, st) => {
    if (err || st.isDirectory()) {
      res.writeHead(404, { 'Content-Type': 'text/plain; charset=utf-8' });
      res.end('404 ' + rel);
      return;
    }
    res.writeHead(200, {
      'Content-Type': MIME[path.extname(file).toLowerCase()] ?? 'application/octet-stream',
      'Content-Length': st.size,
    });
    fs.createReadStream(file).pipe(res);
  });
}

// -- server -----------------------------------------------------------

function log(msg) {
  process.stdout.write(`${new Date().toISOString().slice(11, 19)}  ${msg}\n`);
}

const server = http.createServer((req, res) => {
  const url = new URL(req.url, 'http://localhost');
  const isRelay = url.pathname === '/api/solat' || url.pathname === '/api/jakim';

  if (req.method === 'OPTIONS' && isRelay) {
    res.writeHead(204, {
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Methods': 'GET, OPTIONS',
    }).end();
    return;
  }
  if (req.method !== 'GET') {
    res.writeHead(405).end('Method Not Allowed');
    return;
  }

  if (isRelay) relay(req, res, url.pathname, url.searchParams);
  else serveStatic(req, res, url.pathname);
});

server.listen(PORT, HOST, () => {
  log(`serving ${ROOT} on http://${HOST}:${PORT}`);
  log('relay: /api/solat?zone={zone}&year={year}   (solat.my)');
  log('relay: /api/jakim?zone={zone}&period={period}   (JAKIM e-Solat)');
  log('put one of those paths in the page’s relay field, then leave this running.');
});
